File: claude/node/fiassed/services/noteService.js
'use strict';

const { Op } = require('sequelize');
const { Note, User, Rating, Attachment, ShareLink, sequelize } = require('../models');
const constants = require('../config/constants');
const auditService = require('./auditService');
const fileService = require('./fileService');

/**
 * Create a new note.
 * ownerId is always taken from the authenticated session, never from request body.
 * @param {string} ownerId - Authenticated user UUID
 * @param {{ title: string, content: string, visibility: string }} data
 * @param {string} [correlationId]
 * @param {string} [ip]
 * @returns {Promise<Note>}
 */
async function createNote(ownerId, data, correlationId, ip) {
  // Integrity: visibility validated against allowlist
  const visibility = data.visibility === 'public' ? 'public' : 'private';

  const note = await Note.create({
    title: data.title,
    content: data.content,
    visibility,
    ownerId
  });

  await auditService.log('note.created', {
    actorId: ownerId,
    targetId: note.id,
    targetType: 'note',
    outcome: 'success',
    metadata: { visibility },
    ip,
    correlationId
  });

  return note;
}

/**
 * Retrieve a note by ID, enforcing visibility and ownership rules.
 * Returns null (not throws) for unauthorized access to prevent enumeration.
 * @param {string} id
 * @param {string|null} requestingUserId
 * @returns {Promise<Note|null>}
 */
async function getNoteById(id, requestingUserId) {
  const note = await Note.findByPk(id, {
    include: [
      { model: User, as: 'owner', attributes: ['id', 'username'] },
      { model: Attachment, as: 'attachments', attributes: ['id', 'originalFilename', 'storedFilename', 'mimeType', 'size', 'createdAt'] }
    ]
  });

  if (!note) return null;

  // Visibility enforced server-side on every access
  if (note.visibility === 'private') {
    if (!requestingUserId || note.ownerId !== requestingUserId) {
      return null;
    }
  }

  return note;
}

/**
 * List notes owned by a user with pagination.
 * @param {string} ownerId
 * @param {number} page
 * @param {number} limit
 * @returns {Promise<{ rows: Note[], count: number }>}
 */
async function listUserNotes(ownerId, page = 1, limit = 10) {
  const offset = (page - 1) * limit;
  const safeLimit = Math.min(limit, constants.PAGINATION.MAX_PAGE_SIZE);

  return Note.findAndCountAll({
    where: { ownerId },
    order: [['updatedAt', 'DESC']],
    limit: safeLimit,
    offset
  });
}

/**
 * Update a note after re-verifying ownership on every call.
 * @param {string} id
 * @param {string} requestingUserId
 * @param {{ title?: string, content?: string, visibility?: string }} data
 * @param {string} [correlationId]
 * @param {string} [ip]
 * @returns {Promise<Note|null>}
 */
async function updateNote(id, requestingUserId, data, correlationId, ip) {
  // Integrity: ownership re-verified from DB on every update
  const note = await Note.findByPk(id);
  if (!note || note.ownerId !== requestingUserId) return null;

  const updates = {};
  if (data.title !== undefined) updates.title = data.title;
  if (data.content !== undefined) updates.content = data.content;
  if (data.visibility !== undefined) {
    updates.visibility = data.visibility === 'public' ? 'public' : 'private';
  }

  await note.update(updates);

  await auditService.log('note.updated', {
    actorId: requestingUserId,
    targetId: id,
    targetType: 'note',
    outcome: 'success',
    ip,
    correlationId
  });

  return note;
}

/**
 * Delete a note with cascading cleanup: shareLinks -> attachments -> ratings -> note.
 * Admins may delete any note; regular users only their own.
 * @param {string} id
 * @param {string} requestingUserId
 * @param {string} requestingUserRole
 * @param {string} [correlationId]
 * @param {string} [ip]
 * @returns {Promise<boolean>}
 */
async function deleteNote(id, requestingUserId, requestingUserRole, correlationId, ip) {
  const note = await Note.findByPk(id, {
    include: [{ model: Attachment, as: 'attachments' }]
  });

  if (!note) return false;

  const isOwner = note.ownerId === requestingUserId;
  const isAdmin = requestingUserRole === constants.ROLES.ADMIN;

  if (!isOwner && !isAdmin) return false;

  // Cascade delete in transaction for atomicity
  await sequelize.transaction(async (t) => {
    await ShareLink.destroy({ where: { noteId: id }, transaction: t });

    // Delete attachment files from disk before DB records
    for (const attachment of note.attachments || []) {
      await fileService.deleteFileFromDisk(attachment.storedFilename);
    }
    await Attachment.destroy({ where: { noteId: id }, transaction: t });

    await Rating.destroy({ where: { noteId: id }, transaction: t });
    await note.destroy({ transaction: t });
  });

  await auditService.log('note.deleted', {
    actorId: requestingUserId,
    targetId: id,
    targetType: 'note',
    outcome: 'success',
    metadata: { deletedBy: isAdmin && !isOwner ? 'admin' : 'owner' },
    ip,
    correlationId
  });

  return true;
}

/**
 * Search notes the requesting user can access.
 * Only their own notes + public notes, parameterized query.
 * @param {string} query
 * @param {string} requestingUserId
 * @param {number} page
 * @returns {Promise<{ rows: Note[], count: number }>}
 */
async function searchNotes(query, requestingUserId, page = 1) {
  const limit = constants.PAGINATION.DEFAULT_PAGE_SIZE;
  const offset = (page - 1) * limit;

  const sanitizedQuery = `%${query}%`;

  return Note.findAndCountAll({
    where: {
      [Op.and]: [
        {
          [Op.or]: [
            { title: { [Op.like]: sanitizedQuery } },
            { content: { [Op.like]: sanitizedQuery } }
          ]
        },
        {
          [Op.or]: [
            { visibility: 'public' },
            { ownerId: requestingUserId }
          ]
        }
      ]
    },
    include: [{ model: User, as: 'owner', attributes: ['username'] }],
    order: [['updatedAt', 'DESC']],
    limit,
    offset
  });
}

/**
 * Get top-rated public notes with at least MIN_RATERS_FOR_TOP distinct raters.
 * @param {number} page
 * @param {number} limit
 * @returns {Promise<Note[]>}
 */
async function getTopRated(page = 1, limit = 10) {
  const offset = (page - 1) * limit;

  // Subquery: find noteIds with at least 3 distinct raters
  const qualifiedNoteIds = await Rating.findAll({
    attributes: ['noteId'],
    group: ['noteId'],
    having: sequelize.where(
      sequelize.fn('COUNT', sequelize.fn('DISTINCT', sequelize.col('userId'))),
      { [Op.gte]: constants.RATINGS.MIN_RATERS_FOR_TOP }
    ),
    raw: true
  });

  const noteIds = qualifiedNoteIds.map((r) => r.noteId);
  if (noteIds.length === 0) return [];

  const notes = await Note.findAll({
    where: {
      id: { [Op.in]: noteIds },
      visibility: 'public'
    },
    include: [
      { model: User, as: 'owner', attributes: ['username'] },
      {
        model: Rating,
        as: 'ratings',
        attributes: []
      }
    ],
    attributes: {
      include: [
        [
          sequelize.fn('AVG', sequelize.col('ratings.value')),
          'avgRating'
        ],
        [
          sequelize.fn('COUNT', sequelize.col('ratings.id')),
          'ratingCount'
        ]
      ]
    },
    group: ['Note.id', 'owner.id'],
    order: [[sequelize.literal('avgRating'), 'DESC']],
    limit,
    offset,
    subQuery: false
  });

  return notes;
}

/**
 * Transfer note ownership atomically with audit log.
 * Admin-only operation.
 * @param {string} noteId
 * @param {string} newOwnerId
 * @param {string} adminId
 * @param {string} [correlationId]
 * @param {string} [ip]
 * @returns {Promise<boolean>}
 */
async function transferOwnership(noteId, newOwnerId, adminId, correlationId, ip) {
  const note = await Note.findByPk(noteId);
  if (!note) return false;

  const newOwner = await User.findByPk(newOwnerId);
  if (!newOwner) return false;

  const previousOwnerId = note.ownerId;

  await sequelize.transaction(async (t) => {
    await note.update({ ownerId: newOwnerId }, { transaction: t });
    await auditService.log('note.ownership_transferred', {
      actorId: adminId,
      targetId: noteId,
      targetType: 'note',
      outcome: 'success',
      metadata: { previousOwnerId, newOwnerId },
      ip,
      correlationId
    });
  });

  return true;
}

module.exports = {
  createNote,
  getNoteById,
  listUserNotes,
  updateNote,
  deleteNote,
  searchNotes,
  getTopRated,
  transferOwnership
};

File: claude/jsp/fiassed/src/main/java/com/loosenotes/service/NoteService.java
package com.loosenotes.service;

import com.loosenotes.dao.NoteDao;
import com.loosenotes.model.AuditEvent;
import com.loosenotes.model.Note;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.List;

/**
 * Application-layer service for note lifecycle management.
 *
 * <p><strong>Security model enforced here (never in DAOs/servlets):</strong>
 * <ul>
 *   <li>Ownership is bound server-side from the authenticated session, never
 *       from request parameters (prevents IDOR / mass-assignment).</li>
 *   <li>Visibility enforcement: private notes are accessible only to their
 *       owner; PUBLIC notes are readable by anyone with a valid session.</li>
 *   <li>Every mutation emits a structured audit event via {@link AuditService}.</li>
 *   <li>{@code updatedAt} timestamps are always set server-side.</li>
 * </ul>
 */
public class NoteService {

    private static final Logger log = LoggerFactory.getLogger(NoteService.class);

    private static final int MAX_TITLE_LENGTH = 200;
    private static final int MAX_CONTENT_LENGTH = 65_535;

    private final NoteDao noteDao;
    private final AuditService auditService;

    /**
     * @param noteDao      DAO for note persistence; must not be {@code null}
     * @param auditService audit sink; must not be {@code null}
     */
    public NoteService(NoteDao noteDao, AuditService auditService) {
        if (noteDao == null) throw new IllegalArgumentException("noteDao must not be null");
        if (auditService == null) throw new IllegalArgumentException("auditService must not be null");
        this.noteDao = noteDao;
        this.auditService = auditService;
    }

    // =========================================================================
    // Create
    // =========================================================================

    /**
     * Creates a new note owned by {@code userId}.
     *
     * <p>The {@code userId} comes from the authenticated session — never from
     * the request body — so clients cannot forge ownership.
     *
     * @param userId     authenticated owner's ID (server-owned; not from request body)
     * @param title      note title (1–200 chars)
     * @param content    note body (up to 65 535 chars)
     * @param visibility PUBLIC or PRIVATE
     * @return the persisted {@link Note} with populated ID and timestamps
     * @throws ServiceException VALIDATION on bad input
     */
    public Note createNote(Long userId, String title, String content, Note.Visibility visibility)
            throws ServiceException {

        validateNoteInput(title, content);

        Note note = new Note();
        note.setUserId(userId);          // server-owned: never accept from request
        note.setTitle(title.trim());
        note.setContent(content);
        note.setVisibility(visibility != null ? visibility : Note.Visibility.PRIVATE);

        Instant now = Instant.now();
        note.setCreatedAt(now);
        note.setUpdatedAt(now);

        Note created = noteDao.create(note);

        auditService.recordEvent(AuditEvent.builder()
                .action("NOTE_CREATED")
                .subjectId(String.valueOf(userId))
                .objectId(String.valueOf(created.getId()))
                .metadata("visibility=" + created.getVisibility())
                .outcome("SUCCESS")
                .build());

        log.info("Note created. noteId={} userId={} visibility={}", created.getId(), userId, created.getVisibility());
        return created;
    }

    // =========================================================================
    // Read
    // =========================================================================

    /**
     * Returns a note if the requesting user is authorised to read it.
     *
     * <p>Access rules:
     * <ul>
     *   <li>Owner may always read their own note.</li>
     *   <li>PUBLIC notes are readable by any authenticated user
     *       ({@code requestingUserId} non-null) and by unauthenticated share-link
     *       visitors ({@code requestingUserId} may be {@code null}).</li>
     *   <li>PRIVATE notes return an ACCESS_DENIED {@link ServiceException} for
     *       non-owners.</li>
     * </ul>
     *
     * @param noteId           note primary key
     * @param requestingUserId authenticated user ID, or {@code null} for anonymous
     * @return the {@link Note}
     * @throws ServiceException NOT_FOUND or ACCESS_DENIED
     */
    public Note getNote(Long noteId, Long requestingUserId) throws ServiceException {
        Note note = requireNote(noteId);

        boolean isOwner = requestingUserId != null && requestingUserId.equals(note.getUserId());

        if (!isOwner && note.getVisibility() == Note.Visibility.PRIVATE) {
            log.warn("ACCESS_DENIED: private note access attempted. noteId={} requestingUserId={}",
                    noteId, requestingUserId);
            throw new ServiceException("ACCESS_DENIED", "You do not have permission to view this note.");
        }

        return note;
    }

    /**
     * Returns a note for editing — ownership is required.
     *
     * @param noteId           note primary key
     * @param requestingUserId authenticated user (must be owner)
     * @return the {@link Note}
     * @throws ServiceException NOT_FOUND or ACCESS_DENIED
     */
    public Note getNoteForEdit(Long noteId, Long requestingUserId) throws ServiceException {
        Note note = requireNote(noteId);
        requireOwnership(note, requestingUserId, "edit");
        return note;
    }

    // =========================================================================
    // Update
    // =========================================================================

    /**
     * Updates an existing note's mutable fields.
     *
     * <p>Ownership is verified server-side.  The {@code updatedAt} timestamp is
     * always set to {@code Instant.now()} — never accepted from the caller.
     * Visibility changes are specifically logged for auditability.
     *
     * @param noteId     note to update
     * @param userId     authenticated user (must be owner)
     * @param title      new title
     * @param content    new content
     * @param visibility new visibility
     * @return {@code true} if the DAO reported a row was updated
     * @throws ServiceException VALIDATION, ACCESS_DENIED, or NOT_FOUND
     */
    public boolean updateNote(Long noteId, Long userId, String title, String content,
                               Note.Visibility visibility) throws ServiceException {

        validateNoteInput(title, content);

        Note existing = requireNote(noteId);
        requireOwnership(existing, userId, "update");

        // Detect visibility change for targeted audit logging.
        boolean visibilityChanged = visibility != null
                && !visibility.equals(existing.getVisibility());

        existing.setTitle(title.trim());
        existing.setContent(content);
        existing.setVisibility(visibility != null ? visibility : existing.getVisibility());
        existing.setUpdatedAt(Instant.now());   // server-set; never from request

        boolean updated = noteDao.update(existing);

        if (updated) {
            AuditEvent.Builder auditBuilder = AuditEvent.builder()
                    .action("NOTE_UPDATED")
                    .subjectId(String.valueOf(userId))
                    .objectId(String.valueOf(noteId))
                    .outcome("SUCCESS");

            if (visibilityChanged) {
                auditBuilder.metadata("visibilityChanged="
                        + existing.getVisibility() + "->" + visibility);
            }

            auditService.recordEvent(auditBuilder.build());
            log.info("Note updated. noteId={} userId={} visibilityChanged={}", noteId, userId, visibilityChanged);
        }

        return updated;
    }

    // =========================================================================
    // Delete
    // =========================================================================

    /**
     * Deletes a note.
     *
     * <p>Authorised when: {@code userId} is the note owner, OR {@code isAdmin}
     * is {@code true}.  Cascading deletes (attachments, ratings, share-links)
     * are handled by FK constraints in the database schema.
     *
     * @param noteId  note to delete
     * @param userId  requesting user's ID
     * @param isAdmin {@code true} if the requesting user holds the ADMIN role
     * @return {@code true} if deleted
     * @throws ServiceException NOT_FOUND or ACCESS_DENIED
     */
    public boolean deleteNote(Long noteId, Long userId, boolean isAdmin) throws ServiceException {
        Note note = requireNote(noteId);

        boolean isOwner = userId != null && userId.equals(note.getUserId());
        if (!isOwner && !isAdmin) {
            log.warn("ACCESS_DENIED: delete attempted by non-owner. noteId={} requestingUserId={}", noteId, userId);
            throw new ServiceException("ACCESS_DENIED", "You do not have permission to delete this note.");
        }

        boolean deleted = noteDao.delete(noteId);

        if (deleted) {
            auditService.recordEvent(AuditEvent.builder()
                    .action("NOTE_DELETED")
                    .subjectId(String.valueOf(userId))
                    .objectId(String.valueOf(noteId))
                    .metadata("byAdmin=" + isAdmin)
                    .outcome("SUCCESS")
                    .build());
            log.info("Note deleted. noteId={} byUserId={} byAdmin={}", noteId, userId, isAdmin);
        }

        return deleted;
    }

    // =========================================================================
    // Listings / search
    // =========================================================================

    /**
     * Returns a page of notes owned by {@code userId}.
     *
     * @param userId   owner's ID
     * @param page     0-based page index
     * @param pageSize records per page
     */
    public List<Note> getUserNotes(Long userId, int page, int pageSize) {
        return noteDao.findByUserId(userId, page, pageSize);
    }

    /**
     * Full-text note search, respecting visibility.
     *
     * <p>Visibility filtering is delegated to the DAO query so that PRIVATE
     * notes belonging to other users are never returned to {@code requestingUserId}.
     *
     * @param requestingUserId authenticated user; DAO uses this to include own
     *                         private notes in results
     * @param query            search term
     * @param page             0-based page index
     * @param pageSize         records per page
     */
    public List<Note> searchNotes(Long requestingUserId, String query, int page, int pageSize) {
        return noteDao.search(requestingUserId, query, page, pageSize);
    }

    /**
     * Returns the top-rated public notes.
     *
     * @param minRatingCount minimum number of ratings a note must have
     * @param pageSize       maximum results to return
     */
    public List<Note> getTopRated(int minRatingCount, int pageSize) {
        return noteDao.findTopRated(minRatingCount, pageSize);
    }

    // =========================================================================
    // Admin operations
    // =========================================================================

    /**
     * Reassigns note ownership to a different user (admin operation).
     *
     * <p>Validates that:
     * <ul>
     *   <li>The target user exists.</li>
     *   <li>The new owner differs from the current owner (no-op guard).</li>
     * </ul>
     *
     * @param noteId        note to reassign
     * @param newOwnerId    target user's ID
     * @param adminId       performing admin's ID (for audit)
     * @param adminUsername performing admin's username (for audit readability)
     * @return {@code true} if the DAO reported a row was updated
     * @throws ServiceException NOT_FOUND or VALIDATION
     */
    public boolean reassignNote(Long noteId, Long newOwnerId, Long adminId, String adminUsername)
            throws ServiceException {

        Note note = requireNote(noteId);

        if (newOwnerId == null) {
            throw new ServiceException("VALIDATION", "New owner ID must not be null.");
        }
        if (newOwnerId.equals(note.getUserId())) {
            throw new ServiceException("VALIDATION", "Note is already owned by that user.");
        }

        Long previousOwnerId = note.getUserId();
        note.setUserId(newOwnerId);
        note.setUpdatedAt(Instant.now());

        boolean updated = noteDao.update(note);

        if (updated) {
            auditService.recordEvent(AuditEvent.builder()
                    .action("NOTE_REASSIGNED")
                    .subjectId(String.valueOf(adminId))
                    .subjectUsername(adminUsername)
                    .objectId(String.valueOf(noteId))
                    .metadata("previousOwnerId=" + previousOwnerId + " newOwnerId=" + newOwnerId)
                    .outcome("SUCCESS")
                    .build());
            log.info("Note reassigned. noteId={} previousOwner={} newOwner={} byAdmin={}",
                    noteId, previousOwnerId, newOwnerId, adminId);
        }

        return updated;
    }

    // =========================================================================
    // Private helpers
    // =========================================================================

    private void validateNoteInput(String title, String content) throws ServiceException {
        if (title == null || title.trim().isEmpty()) {
            throw new ServiceException("VALIDATION", "Note title must not be empty.");
        }
        if (title.length() > MAX_TITLE_LENGTH) {
            throw new ServiceException("VALIDATION",
                    "Note title must not exceed " + MAX_TITLE_LENGTH + " characters.");
        }
        if (content == null) {
            throw new ServiceException("VALIDATION", "Note content must not be null.");
        }
        if (content.length() > MAX_CONTENT_LENGTH) {
            throw new ServiceException("VALIDATION",
                    "Note content must not exceed " + MAX_CONTENT_LENGTH + " characters.");
        }
    }

    private Note requireNote(Long noteId) throws ServiceException {
        if (noteId == null) {
            throw new ServiceException("VALIDATION", "Note ID must not be null.");
        }
        Note note = noteDao.findById(noteId);
        if (note == null) {
            throw new ServiceException("NOT_FOUND", "Note not found.");
        }
        return note;
    }

    private void requireOwnership(Note note, Long userId, String operation) throws ServiceException {
        if (userId == null || !userId.equals(note.getUserId())) {
            log.warn("ACCESS_DENIED: {} attempted by non-owner. noteId={} requestingUserId={}",
                    operation, note.getId(), userId);
            throw new ServiceException("ACCESS_DENIED",
                    "You do not have permission to " + operation + " this note.");
        }
    }
}
